nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java

```diff
@@ -0,0 +1,414 @@
+                            .valid(false)
+                            .explanation("Authentication protocol and password must be set when using authNoPriv security level.")
+                            .build());
+                }
+                break;
+            case "authPriv":
+                if(!isAuthProtOK || !isAuthPwdSet || !isPrivProtOK || !isPrivPwdSet) {
+                    problems.add(new ValidationResult.Builder()
+                            .input("SNMP Security Level")
+                            .valid(false)
+                            .explanation("All protocols and passwords must be set when using authPriv security level.")
+                            .build());
+                }
+                break;
+            case "noAuthNoPriv":
+            default:
+                break;
+            }
+        } else {
+            final boolean isCommunitySet = validationContext.getProperty(SNMP_COMMUNITY).isSet();
+            if(!isCommunitySet) {
+                problems.add(new ValidationResult.Builder()
+                        .input("SNMP Community")
+                        .valid(false)
+                        .explanation("SNMP Community must be set with SNMPv1 and SNMPv2c.")
+                        .build());
+            }
+        }
+
+        return problems;
+    }
+
+    /**
+     * Delegate method to supplement
+     * {@link #onTrigger(ProcessContext, ProcessSession)}. It is implemented by
+     * sub-classes to perform {@link Processor} specific functionality.
+     *
+     * @param context
+     *            instance of {@link ProcessContext}
+     * @param session
+     *            instance of {@link ProcessSession}
+     * @throws ProcessException Process exception
+     */
+    protected abstract void onTriggerSnmp(ProcessContext context, ProcessSession session) throws ProcessException;
+
+    /**
+     * Delegate method to supplement building of target {@link SNMPWorker} (see
+     * {@link SNMPSetter} or {@link SNMPGetter}) and is implemented by
+     * sub-classes.
+     *
+     * @param context
+     *            instance of {@link ProcessContext}
+     * @return new instance of {@link SNMPWorker}
+     */
+    protected abstract T finishBuildingTargetResource(ProcessContext context);
+
+    /**
+     * Builds target resource.
+     * @param context Process context
+     */
+    private void buildTargetResource(ProcessContext context) {
+        if((this.transportMapping == null) || !this.transportMapping.isListening() || (this.snmp == null)) {
+            try {
+                this.transportMapping = new DefaultUdpTransportMapping();
+                this.snmp = new Snmp(this.transportMapping);
+
+                if("SNMPv3".equals(context.getProperty(SNMP_VERSION).getValue())) {
+                    USM usm = new USM(SecurityProtocols.getInstance(), new OctetString(MPv3.createLocalEngineID()), 0);
+                    SecurityModels.getInstance().addSecurityModel(usm);
```

Comment:
I just initiated a DISCUSS thread on the dev list. Basically while there is not much we can do at the moment, the above code will set the security model that will be applicable across all instances of SNMP processors within a single instance of NiFi.